### 1. Clasificación del dataset make_moons con Keras


In [4]:
import numpy as np
import matplotlib.pyplot as plt

from sklearn.datasets import make_moons
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score

import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.layers import Input

X, y = make_moons(n_samples=1000, noise=0.2, random_state=42)

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

model_moons = Sequential([
    Input(shape=(2,)),
    Dense(16, activation="relu", input_shape=(2,)),
    Dense(16, activation="relu"),
    Dense(1, activation="sigmoid")
])

model_moons.compile(
    optimizer=Adam(learning_rate=0.01),
    loss="binary_crossentropy",
    metrics=["accuracy"]
)

history_moons = model_moons.fit(
    X_train, y_train,
    epochs=100,
    batch_size=32,
    validation_split=0.2,
    verbose=0
)

test_loss_moons, test_acc_moons = model_moons.evaluate(X_test, y_test, verbose=0)
test_acc_moons


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


0.9649999737739563

### 2. Clasificación del dataset load_digits con Keras


In [6]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score
from tensorflow.keras.utils import to_categorical
from sklearn.datasets import load_digits

digits = load_digits(n_class=6)
X, y = digits.data, digits.target
n_samples, n_features = X.shape


X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

y_train_cat = to_categorical(y_train)
y_test_cat = to_categorical(y_test)

model_digits = Sequential([
    Dense(32, activation="relu", input_shape=(n_features,)),
    Dense(32, activation="relu"),
    Dense(6, activation="softmax")
])

model_digits.compile(
    optimizer=Adam(learning_rate=0.001),
    loss="categorical_crossentropy",
    metrics=["accuracy"]
)

history_digits = model_digits.fit(
    X_train, y_train_cat,
    epochs=50,
    batch_size=32,
    validation_split=0.2,
    verbose=0
)

test_loss_digits, test_acc_digits = model_digits.evaluate(X_test, y_test_cat, verbose=0)
test_acc_digits


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


0.981566846370697

En el dataset make_moons el modelo de red neuronal implementado con Keras alcanzó una exactitud cercana al 96 %, evidenciando su capacidad para modelar fronteras de decisión no lineales. Para el dataset load_digits (6 clases), la red obtuvo una exactitud aproximada del 98 %, mostrando un desempeño sobresaliente en la clasificación multiclase. Estos resultados confirman la eficacia de Keras para entrenar redes neuronales de forma eficiente y estable, en contraste con la implementación manual desarrollada en el laboratorio anterior.